feat: Add docs about request isolation in Node SDK (#11378)
getsentry/sentry-docs: docs/platforms/javascript/common/enriching-events/process-isolation/index.mdx

---
title: Process Isolation
description: "Learn more about how process isolation (or request isolation) works in the Sentry SDK."
supported:
  - javascript.nextjs
  - javascript.node
  - javascript.connect
  - javascript.express
  - javascript.fastify
  - javascript.hapi
  - javascript.koa
  - javascript.nestjs
  - javascript.nuxt
  - javascript.solidstart
  - javascript.sveltekit
  - javascript.astro
  - javascript.remix
notSupported:
  - javascript
---

In server-side environments, the <PlatformLink to='/enriching-events/scopes'>isolation scope</PlatformLink> is automatically forked around request boundaries. This means that each request will have its own isolation scope, and data set on the isolation scope will only apply to events captured during that request. This is done automatically by the SDK.

> However, there are also other cases where you may want to have isolation, for example in background jobs or when you want to isolate a specific part of your code. In these cases, you can use `Sentry.withIsolationScope()` to create a new isolation scope that is valid inside of the callback you pass to it - see [Using withIsolationScope](../scopes/#using-withisolationscope).
> 
> The following example shows how you can use `withIsolationScope` to attach data for a specific job run:

Author: Maybe slightly different:

```suggestion
However, there are also other times when you may want to have isolation, for example, in background jobs or when you want to isolate a specific part of your code. In these cases, you can use `Sentry.withIsolationScope()` to create a new isolation scope that's valid inside of the callback you pass to it. Learn more about using [withIsolationScope](../scopes/#using-withisolationscope).

The following example shows how you can use `withIsolationScope` to attach data to a specific job run:
```

---

> title: Process Isolation

Reviewer: m: I'm not sure if "Process" isolation is a good term for the heading. We worry about isolating certain jobs/tasks/requests _within_ one process. 

I'd rather call this page "Request Isolation" or more generally "Task Isolation". WDYT?

Author: Yeah, I was also not 100% sure about this. My first instinct was also to go with `Request Isolation` but this does not really cover e.g. isolation background jobs 🤔 but maybe request isolation is still more understandable 🤔

Reviewer 2: I vote for "Request Isolation". I think it's most clear and I think that it's ok that it doesn't cover everything. "Data isolation" is super broad and can mean just about anything. Same thing with "Process isolation".

Author: will rename it!

description: "Learn more about how process isolation (or request isolation) works in the Sentry SDK."
supported:
  - javascript.nextjs
  - javascript.node
  - javascript.connect
  - javascript.express
  - javascript.fastify
  - javascript.hapi
  - javascript.koa
  - javascript.nestjs
  - javascript.nuxt
  - javascript.solidstart
  - javascript.sveltekit
  - javascript.astro
  - javascript.remix
notSupported:
  - javascript
---

In server-side environments, the <PlatformLink to='/enriching-events/scopes'>isolation scope</PlatformLink> is automatically forked around request boundaries. This means that each request will have its own isolation scope, and data set on the isolation scope will only apply to events captured during that request. This is done automatically by the SDK.

> However, the request isolation happens when the request itself is processed. This means that if you e.g. have a middleware where you want to set Sentry data (e.g. `Sentry.setUser()` in an auth middleware), you have to manually fork the isolation scope with `Sentry.withIsolationScope()` - see [Using withIsolationScope](../scopes/#using-withisolationscope).

Reviewer: not actionable for docs but
that's a massive limitation actually 😬 I was under the impression that the OTEL http instrumentation would fork the isolation scope automatically before any express middleware is evaluated. Is that not the case?

Reviewer 3: In my app, this _is_ the case with the "standard" OTel instrumentation packages. I can't speak to Sentry's internals, but it's very very strange that this doesn't just work out of the box.

Regardless of that: the language "when the request is processed" is, to me, imprecise and maybe even inaccurate. I consider middleware execution to very much be part of "request processing" - after all, I have middleware parsing request bodies, performing authorization checks, redirecting authenticated requests to the login page, and so on. My app has many requests that are completely handled by middleware!

Author: so in otel you do not have this concept of scopes and capturing data on them for a request etc., so this problem does not really exist there, as far as I can tell.

We have the problem that the way that the auto-isolation works is actually forking the isolation scope after a http.server span is emitted. And this is of course only emitted once the request is actually "done" (or at least started), not when the middlewares run. I agree that this kind of sucks and we should see if we find a way to solve this better. I'll adjust the wording for this for now!

Author: OK, I did some more digging - actually this is not correct! Middlewares _are_ actually correctly isolated in express, so I will remove this section again (the general content is still good). 

@nwalters512 the problem in your app is not a fundamental problem with middlewares, but it happens because you added http instrumentation yourself, which overwrites ours (and this is what handles the auto isolation). See https://docs.sentry.io/platforms/javascript/guides/node/opentelemetry/custom-setup/#custom-http-instrumentation for details on this.  Sentry relies on the http instrumentation for a bunch of things. It would be nice to find a way to warn about this, though 🤔

Reviewer 4: Hmmm.. if we can't create an instance of of OTEL http instrumentation, how are we supposed to go about registering it with our existing OTEL setup?

You're suggesting:

```ts
// do not import/create a new HttpInstrumentation instance
// import { HttpInstrumentation } from '@opentelemetry/instrumentation-http';
const sentryHttpIntegration = Sentry.httpIntegration();

const sentryClient = Sentry.init({
    dsn: process.env.SENTRY_DSN,
    integrations: (defaultIntegration) =>
        defaultIntegration.map((integration) => (integration.name === 'Http' ? sentryHttpIntegration : integration)),
    release,
    environment,
    skipOpenTelemetrySetup: true,
    tracesSampleRate: 1.0,
});
```

if that's the case what we're unable to [register the instrumentation](https://github.com/open-telemetry/opentelemetry-js/blob/main/packages/opentelemetry-sdk-trace-node/README.md#instrumentation-configuration)

Author: You don't need to add the http instrumentation yourself at all - just do not run `registerInstrumentations({ instrumentation: [new HttpInstrumentation()] })`, this happens automatically under the hood. Does that lead to problems/does not work as expected for you, with your custom setup? 🤔 We could expose our HttpInstrumentation and allow you to add this yourself, I guess, if that's necessary!

Reviewer 4: I updated my demo sentry + OTEL setup as follows:

```ts
import * as Sentry from '@sentry/node';
import { SentryPropagator, SentrySampler } from '@sentry/opentelemetry';
// import { registerInstrumentations } from '@opentelemetry/instrumentation';
// import { ExpressInstrumentation } from '@opentelemetry/instrumentation-express';
// import { HttpInstrumentation } from '@opentelemetry/instrumentation-http';
import { BatchSpanProcessor } from '@opentelemetry/sdk-trace-base';
import { NodeTracerProvider } from '@opentelemetry/sdk-trace-node';

const release = 'sentry-scope-testing';
const environment = 'local';

const sentryClient = Sentry.init({
    dsn: process.env.SENTRY_DSN,
    beforeSend: (event, hint, ...args) => {
        const { contexts, exception, extra, tags, message, user } = event;
        console.dir(
            {
                whoami: 'sentry:beforeSend',
                event: { contexts, exception, extra, tags, message, user },
                hint,
                args,
            },
            { depth: null }
        );
        return event;
    },
    release,
    environment,
    skipOpenTelemetrySetup: true,
    tracesSampleRate: 1.0,
});

const provider = new NodeTracerProvider({
    sampler: new SentrySampler(sentryClient!),
});

provider.addSpanProcessor(
    // mock processor to avoid polluting console output
    new BatchSpanProcessor({
        export: (_spans, cb) => cb({ code: 0 }),
        shutdown: () => Promise.resolve(),
        forceFlush: () => Promise.resolve(),
    })
);

provider.register({
    propagator: new SentryPropagator(),
    contextManager: new Sentry.SentryContextManager(),
});

// registerInstrumentations({
//     instrumentations: [new ExpressInstrumentation(), new HttpInstrumentation()],
// });

Sentry.validateOpenTelemetrySetup();

export const initializedSentry = Sentry;
```

Then for the express middleware, i removed forking of the isolation scope to see if Middleware is correctly isolated in express:

```ts
const auth: RequestHandler = (req, res, next) => {
    const authUser = Users.find((u) => u.id === req.headers['authorization']);
    // Sentry.withIsolationScope(() => {
    if (!authUser) {
        Sentry.setTag('Authenticated', false);
        Sentry.setTag('UserID', null);
        Sentry.setUser(null);
        next(new Error('Authentication Error'));
    } else {
        Sentry.setTag('Authenticated', true);
        Sentry.setTag('UserID', authUser.id);
        Sentry.setUser(authUser);
        res.locals.authUser = authUser;
        next();
    }
    // });
};
```

However, my output reverted to showing the issue:

```js
{
  whoami: 'sentry:beforeSend',
  event: { ... },
    exception: { ... },
    extra: {
      expectedUser: { id: '4', email: 'foo4@example.com', name: 'foo example4' }
    },
    tags: { Authenticated: true, UserID: '1' },
    message: undefined,
    user: { id: '1', email: 'foo@example.com', name: 'foo example' }
  },
  hint: { ... },
  args: []
}
```

Reviewer 4: Furthermore, switching my "mock exporter" to the `ConsoleSpanExporter` does not print an OTEL spans

```ts
provider.addSpanProcessor(
    // mock processor to avoid polluting console output
    new BatchSpanProcessor(new ConsoleSpanExporter())
    // new BatchSpanProcessor({
    //     export: (_spans, cb) => cb({ code: 0 }),
    //     shutdown: () => Promise.resolve(),
    //     forceFlush: () => Promise.resolve(),
    // })
);
```

Author: I opened a PR here: https://github.com/sbriceland/sentry-debug/pull/1 where we can discuss this more!

> This is also necessary if you want to isolate a non-request process, e.g. a background job.
> 
> The following example shows how you can use `withIsolationScope` to attach a user and a tag in an auth middleware:
> 
> ```javascript
> const auth = (req, res, next) => {
>   Sentry.withIsolationScope(() => {
>     const authUser = findUserForHeader(req.headers["authorization"]);
>     if (!authUser) {
>       Sentry.setTag("Authenticated", false);
>       Sentry.setUser(null);
>       next(new Error("Authentication Error"));
>     } else {
>       Sentry.setTag("Authenticated", true);
>       Sentry.setUser(authUser);
>       next();
>     }
>   });
> };
> ```
> 
> This way, the user & tag will only be attached to events captured during the request that passed the auth middleware.

Reviewer: l: I'd move the sentence about bg jobs after the user/middleware example for consistency

```suggestion
The following example shows how you can use `withIsolationScope` to attach a user and a tag in an auth middleware:

```javascript
const auth = (req, res, next) => {
  Sentry.withIsolationScope(() => {
    const authUser = findUserForHeader(req.headers["authorization"]);
    if (!authUser) {
      Sentry.setTag("Authenticated", false);
      Sentry.setUser(null);
      next(new Error("Authentication Error"));
    } else {
      Sentry.setTag("Authenticated", true);
      Sentry.setUser(authUser);
      next();
    }
  });
};
```

This way, the user & tag will only be attached to events captured during the request that passed the auth middleware.

Manually calling `withIsolationScope` is also necessary if you want to isolate a non-request task, like background jobs or queue task.
```

Reviewer: argh the nested ``` lines destroy the diff. I just moved the bg job sentence to the bottom and slightly adjusted it.

Author: hmm, I do not want to bury this at the very bottom because it is actually a very important aspect that we don't really cover anywhere 🤔 like, if you have a background job or similar you will _definitely always_ have to do this manually 🤔
